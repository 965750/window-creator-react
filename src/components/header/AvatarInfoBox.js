import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styles from './AvatarInfoBox.module.scss'
import { connect } from 'react-redux'

class AvatarInfoBox extends Component {
    state = {
        showInfoBox: false,
    }

    showInfoBox = value => {
        this.setState({
            showInfoBox: value,
        })
    }

    hideInfoBox = e => {
        if (this.state.showInfoBox && !this.node.contains(e.target)) {
            this.setState({
                showInfoBox: false,
            })
        }
    }

    componentDidMount() {
        window.addEventListener('click', this.hideInfoBox)
    }

    componentWillUnmount() {
        window.removeEventListener('click', this.hideInfoBox)
    }

    render() {
        let infoBox
        let initials = this.props.user.firstName
            ? this.props.user.firstName[0] + this.props.user.lastName[0]
            : ''

        if (this.state.showInfoBox)
            infoBox = (
                <div
                    className={`${styles.infoBox} p-2 absolute right-0 border border-gullGray rounded bg-wildSand`}
                >
                    <p>{`${this.props.user.firstName} ${this.props.user.lastName}`}</p>
                    <p>{this.props.auth.email}</p>
                </div>
            )

        return (
            <div
                className={`${styles.avatar} z-30 mt-md rounded-full bg-first cursor-pointer`}
                onMouseOver={this.showInfoBox}
                ref={n => (this.node = n)}
            >
                <p className="uppercase text-center pt-1 leading-loose text-white text-xl font-bold">
                    {initials}
                </p>
                {infoBox}
            </div>
        )
    }
}

AvatarInfoBox.propTypes = {
    user: PropTypes.shape({
        firstName: PropTypes.string,
        lastName: PropTypes.string,
        isLoaded: PropTypes.bool,
        isEmpty: PropTypes.bool,
    }),
    auth: PropTypes.shape({
        email: PropTypes.string,
        isEmpty: PropTypes.bool,
        lastLoginAt: PropTypes.string,
    }),
}

const mapStateToProps = state => {
    return {
        user: state.firebase.profile,
        auth: state.firebase.auth,
    }
}

export default connect(mapStateToProps)(AvatarInfoBox)
