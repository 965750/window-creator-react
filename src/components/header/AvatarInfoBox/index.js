import { connect } from 'react-redux'
import { FormattedMessage } from 'react-intl'

import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './AvatarInfoBox.module.scss'

class AvatarInfoBox extends Component {
  state = {
    showInfoBox: false,
  }

  componentDidMount() {
    window.addEventListener('click', this.hideInfoBox)
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.hideInfoBox)
  }

  toggleInfoBox = () => {
    this.setState((prevState) => ({
      showInfoBox: !prevState.showInfoBox,
    }))
  }

  hideInfoBox = (e) => {
    if (this.state.showInfoBox && !this.node.contains(e.target)) {
      this.setState({
        showInfoBox: false,
      })
    }
  }

  render() {
    let infoBox
    const initials = this.props.user.firstName
      ? this.props.user.firstName[0] + this.props.user.lastName[0]
      : ''

    if (this.state.showInfoBox) {
      infoBox = (
        <div
          data-testid="avatarInfoBox"
          className={`${styles.infoBox} p-2 absolute right-0 border border-gullGray rounded bg-wildSand`}
        >
          <p>{`${this.props.user.firstName} ${this.props.user.lastName}`}</p>
          <p>{this.props.auth.email}</p>
          <p>
            <span>
              <FormattedMessage
                id="saved projects:"
                defaultMessage="saved projects:"
              />
            </span>
            <span>{` ${this.props.user.windows.length}`}</span>
          </p>
        </div>
      )
    }

    return (
      <div
        data-testid="avatar"
        className={`${styles.avatar} z-30 mt-md rounded-full bg-mountainMeadow cursor-pointer`}
        onClick={this.toggleInfoBox}
        ref={n => (this.node = n)} // eslint-disable-line
      >
        <p className="uppercase text-center pt-1 leading-loose text-white text-xl font-bold">
          {initials}
        </p>
        {infoBox}
      </div>
    )
  }
}

AvatarInfoBox.propTypes = {
  user: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    isLoaded: PropTypes.bool,
    isEmpty: PropTypes.bool,
    windows: PropTypes.array,
  }),
  auth: PropTypes.shape({
    email: PropTypes.string,
    isEmpty: PropTypes.bool,
    lastLoginAt: PropTypes.string,
  }),
}

const mapStateToProps = (state) => ({
  user: state.firebase.profile,
  auth: state.firebase.auth,
})

export default connect(mapStateToProps)(AvatarInfoBox)
