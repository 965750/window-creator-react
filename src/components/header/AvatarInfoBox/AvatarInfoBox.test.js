import '@testing-library/jest-dom/extend-expect'

import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup, fireEvent } from '@testing-library/react'

import AvatarInfoBox from './index'
import messages from '../../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
  },
  firebase: {
    auth: {
      email: 'mockEmail@gmail.com'
    },
    profile: {
      firstName: 'mockName',
      lastName: 'mockLastName',
      isLoaded: true,
      isEmpty: false,
      windows: [{
        id: 'mockId',
        save: {
          name: 'mockWindowName'
        }
      }]
    }
  }
}
const reducer = (state = initialStore, action) => {
  switch (action.type) {
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('AvatarInfoBox', () => {
  let wrapper
  const initialProps = {}

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <IntlProvider
        locale="en"
        messages={messages[store.getState().theme.lang]}
      >
        <AvatarInfoBox {...props} />
      </IntlProvider>,
    )
  }

  afterEach(cleanup)

  it('should match snapshot', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should have informations about user in opened box', () => {
    setUp()

    const avatar = wrapper.getByTestId('avatar')
    fireEvent.click(avatar)

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should compose initials for avatar from user\'s firstName and lastName', () => {
    setUp()

    expect(wrapper.queryByText('mm')).not.toBeNull()
  })

  it('should be able to open box then close it', () => {
    setUp()

    const avatar = wrapper.getByTestId('avatar')

    fireEvent.click(avatar)
    expect(wrapper.queryByTestId('avatarInfoBox')).not.toBeNull()

    fireEvent.click(avatar)
    expect(wrapper.queryByTestId('avatarInfoBox')).toBeNull()
  })
})
