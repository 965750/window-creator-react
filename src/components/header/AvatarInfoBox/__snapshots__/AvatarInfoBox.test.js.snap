// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarInfoBox should have informations about user in opened box 1`] = `
<DocumentFragment>
  <div
    class="avatar z-30 mt-md rounded-full bg-mountainMeadow cursor-pointer"
    data-testid="avatar"
  >
    <p
      class="uppercase text-center pt-1 leading-loose text-white text-xl font-bold"
    >
      mm
    </p>
    <div
      class="infoBox p-2 absolute right-0 border border-gullGray rounded bg-wildSand"
      data-testid="avatarInfoBox"
    >
      <p>
        mockName mockLastName
      </p>
      <p>
        mockEmail@gmail.com
      </p>
      <p>
        <span>
          zapisanych projekt√≥w:
        </span>
        <span>
           1
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AvatarInfoBox should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="avatar z-30 mt-md rounded-full bg-mountainMeadow cursor-pointer"
    data-testid="avatar"
  >
    <p
      class="uppercase text-center pt-1 leading-loose text-white text-xl font-bold"
    >
      mm
    </p>
  </div>
</DocumentFragment>
`;
